redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+    def __init__(self, predicate, child, *, input_info,
+                 full_event=False, **kwargs):
+        """Initialize the node
+
+        Parameters
+        ----------
+        predicate: callable
+            The function which returns True if the event is to propagate
+            further in the pipeline
+        child: EventStream instance
+            The source of the data
+        input_info: dict
+            describe the incoming streams
+        full_event: bool, optional
+            If True expose the full event dict to the predicate, if False
+            only expose the data from the event
+        """
+        self.predicate = predicate
+
+        EventStream.__init__(self, child, input_info=input_info, **kwargs)
+        self.full_event = full_event
+        self.generate_provenance(predicate)
+
+    def event(self, doc):
+        if not self.full_event:
+            g = self.event_guts(doc)
+        else:
+            g = doc
+        if self.predicate(g):
+            return super().event(doc[0])
+
+
+class accumulate(EventStream):
+    """Accumulate results with previous state
+
+    This preforms running or cumulative reductions, applying the function
+    to the previous total and the new element.  The function should take
+    two arguments, the previous accumulated state and the next element and
+    it should return a new accumulated state.
+    """
+    def __init__(self, func, child, state_key=None, *,
+                 output_info=None,
+                 input_info=None, start=no_default):
+        """Initialize the node
+
+        Parameters
+        ----------
+        func: callable
+            The function to map on the event data
+        child: EventStream instance
+            The source of the data
+        state_key: str
+            The keyword for current accumulated state in the func
+        input_info: dict
+            describe the incoming streams
+        output_info: list of tuples
+            describe the resulting stream
+        start: any or callable, optional
+            Starting value for accumulation, if no_default use the event data
+            dictionary, if callable run that callable on the event data, else
+            use `start` as the starting data, defaults to no_default
+        """
+        self.state_key = state_key
+        self.func = func
+        self.state = start
+        EventStream.__init__(self, child, input_info=input_info,
+                             output_info=output_info)
+        self.generate_provenance(func)
+
+    def event(self, doc):
+        doc = self.event_guts(doc)
+        # TODO: this handling of the initial state is a bit clunky
+        # I need to decide if state is going to be the array or the dict
+        if self.state is no_default:
+            self.state = doc
+        # in case we need a bit more flexibility eg lambda x: np.empty(x.shape)
+        elif hasattr(self.state, '__call__'):
```

Comment:
The state, which is a callable, will operate on the doc. This will result in something not callable (usually an array) for the first value. For instance, you might want to start with an empty array, but you don't know how big that array is so you need some way to construct an empty array, you can do this with the `lambda x: np.empty(x.shape)`. Then since we are done with our state callable we will reassign the name `self.state` to be what the callable `self.state` returned.